core/blockchain/tokenstate.go

```diff
@@ -0,0 +1,403 @@
+	if bytesRead == 0 {
+		return nil, errDeserialize("unexpected end of data while reading number of balances")
+	}
+	offset += bytesRead
+	if numOfUpdates > 0 {
+		updates = make([]balanceUpdate, numOfUpdates)
+		for i := uint64(0); i < numOfUpdates; i++ {
+			//type
+			updateType, bytesRead := deserializeVLQ(data[offset:])
+			if bytesRead == 0 {
+				return nil, fmt.Errorf("unexpected end of data while reading balance update type at update[%d]", i)
+			}
+			offset += bytesRead
+			//meerAmount
+			meerAmount, bytesRead := deserializeVLQ(data[offset:])
+			if bytesRead == 0 {
+				return nil, fmt.Errorf("unexpected end of data while reading meer amount at update[%d]", i)
+			}
+			offset += bytesRead
+			//tokenId
+			tokenId, bytesRead := deserializeVLQ(data[offset:])
+			if bytesRead == 0 {
+				return nil, fmt.Errorf("unexpected end of data while reading token id at update[%d]", i)
+			}
+			offset += bytesRead
+			//tokenAmount
+			tokenAmount, bytesRead := deserializeVLQ(data[offset:])
+			if bytesRead == 0 {
+				return nil, fmt.Errorf("unexpected end of data while reading token amount at update[%d]", i)
+			}
+			offset += bytesRead
+
+			updates[i] = balanceUpdate{
+				typ:         balanceUpdateType(updateType),
+				meerAmount:  int64(meerAmount),
+				tokenAmount: types.Amount{Value:int64(tokenAmount), Id:types.CoinID(uint16(tokenId))},
+			}
+		}
+	}
+	return &tokenState{balances: balances, updates: updates}, nil
+}
+
+// dbPutTokenState put a token balance record into the token state database.
+// the key is the provided block hash
+func dbPutTokenState(dbTx database.Tx, hash *hash.Hash, ts tokenState) error {
+	// Serialize the current token state.
+	serializedData, err := serializeTokeState(ts)
+	if err != nil {
+		return err
+	}
+	// Store the current token balance record into the token state database.
+	meta := dbTx.Metadata()
+	bucket := meta.Bucket(dbnamespace.TokenBucketName)
+	return bucket.Put(hash[:], serializedData)
+}
+
+// dbFetchTokenState fetch the token balance record from the token state database.
+// the key is the input block hash.
+func dbFetchTokenState(dbTx database.Tx, hash hash.Hash) (*tokenState, error) {
+	// if it is genesis hash, return empty tokenState directly
+	if *params.ActiveNetParams.GenesisHash == hash {
+		return &tokenState{}, nil
+	}
+	// Fetch record from the token state database by block hash
+	meta := dbTx.Metadata()
+	bucket := meta.Bucket(dbnamespace.TokenBucketName)
+	v := bucket.Get(hash[:])
+	if v == nil {
+		return nil, fmt.Errorf("tokenstate db can't find record from block hash : %v", hash)
+	}
+	// deserialize the fetched token state record
+	return deserializeTokenState(v)
+}
+
+const dagConfirmFactor = 10
+
+func (b *BlockChain) calculateTokenBalance(dbTx database.Tx, node *blockNode) (tokenBalances, error) {
+
+	result := tokenBalances{}
+
+	// the passed-in node should always has already been ordered at this level
+	if !node.IsOrdered() {
```

Comment:
The newly added block may not be ordered immediately

Reply:
has removed